libraries/AP_Math/AP_Math.h

```diff
@@ -17,153 +15,135 @@
-}
-
-// constrain a int32_t value
-static inline int32_t constrain_int32(int32_t amt, int32_t low, int32_t high) {
-	return ((amt)<(low)?(low):((amt)>(high)?(high):(amt)));
-}
-
-//matrix algebra
-bool inverse(float x[], float y[], uint16_t dim);
-
-// degrees -> radians
-static inline float radians(float deg) {
-	return deg * DEG_TO_RAD;
-}
-
-// radians -> degrees
-static inline float degrees(float rad) {
-	return rad * RAD_TO_DEG;
-}
-
-// square
-static inline float sq(float v) {
-	return v*v;
-}
-
-// 2D vector length
-static inline float pythagorous2(float a, float b) {
-	return sqrtf(sq(a)+sq(b));
-}
-
-// 3D vector length
-static inline float pythagorous3(float a, float b, float c) {
-	return sqrtf(sq(a)+sq(b)+sq(c));
+template <class T>
+auto wrap_180_cd(const T &angle) -> decltype(wrap_180(angle, 100.f)) {
+    return wrap_180(angle, 100.f);
 }
 
+/* 
+ * @brief: Gets two values and returns the smaller one.
+ */
 template<typename A, typename B>
-static inline auto MIN(const A &one, const B &two) -> decltype(one < two ? one : two) {
+auto MIN(const A &one, const B &two) -> decltype(one < two ? one : two) { 
     return one < two ? one : two;
 }
 
+/* 
+ * @brief: Gets two values and returns the bigger one.
+ */
 template<typename A, typename B>
-static inline auto MAX(const A &one, const B &two) -> decltype(one > two ? one : two) {
+auto MAX(const A &one, const B &two) -> decltype(one > two ? one : two) {  
     return one > two ? one : two;
 }
 
-inline uint32_t hz_to_nsec(uint32_t freq)
+/*
+ * @brief: linear interpolation based on a variable in a range
+ */
+template <class T>
+T linear_interpolate(const T &low_output, const T &high_output,
+                     const T &var_value,
+                     const T &var_low, const T &var_high)
 {
-    return NSEC_PER_SEC / freq;
+    if (var_value <= var_low) {
+        return low_output;
+    }
+    if (var_value >= var_high) {
+        return high_output;
+    }
+    // avoid zero divisions or zero like divisions
+    auto var_diff = var_high - var_low;
+    if(is_zero(var_diff)) {
+        return low_output;
+    }
+    
+    T p = (var_value - var_low) / var_diff;
+    return low_output + p * (high_output - low_output);
 }
 
-inline uint32_t nsec_to_hz(uint32_t nsec)
-{
-    return NSEC_PER_SEC / nsec;
+/*
+ * Converter functions
+ *  - Avoid zero divisions
+ *  - Inheritss a float cast (because of PX4)
+ */
+template<class T>
+T hz_to_nsec(const T &freq) {
+    return freq != 0 ? NSEC_PER_SEC / freq : 0;
 }
 
-inline uint32_t usec_to_nsec(uint32_t usec)
-{
+template<class T>
+T nsec_to_hz(const T &nsec) {
+    return nsec != 0 ? NSEC_PER_SEC / nsec : 0;
+}
+
+template<class T>
+T usec_to_nsec(const T &usec) {
     return usec * NSEC_PER_USEC;
 }
 
-inline uint32_t nsec_to_usec(uint32_t nsec)
-{
-    return nsec / NSEC_PER_USEC;
+template<class T>
+T nsec_to_usec(const T &nsec) {
+    return nsec != 0 ? nsec / NSEC_PER_USEC : 0;
```

Comment:
Em 14 de abr de 2016 5:33 AM, "Daniel Frenzel" <notifications@github.com>
escreveu:
>
> In libraries/AP_Math/AP_Math.h:
>
> >      return usec * NSEC_PER_USEC;
> >  }
> >
> > -inline uint32_t nsec_to_usec(uint32_t nsec)
> > -{
> > -    return nsec / NSEC_PER_USEC;
> > +template<class T>
> > +T nsec_to_usec(const T &nsec) {
> > +    return nsec != 0 ? nsec / NSEC_PER_USEC : 0;
>
> I used a reference in case this function will ever be used for bigger
types like vectors matrices.

I don't think it will ever be. If it is, we can change.

> The zero check is in case of integer types, which would create a program
termination and because this check does not hurt. I guess, you still want
the zero check out. We it then be pushed?

I was talking about this specific case. You are not dividing by it so
there's no div by 0 here.

Lucas De Marchi